scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -1156,4 +1158,46 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
     )
     new MultiMapSideInput[K, V](o)
   }
+
+  /**
+   * Convert this SCollection to a SideInput, mapping key-value pairs of each window to a
+   * `Map[key, value]`, to be used with [[SCollection.withSideInputs]]. It is required that each
+   * key of the input be associated with a single value.
+   *
+   * Currently, the resulting map is required to fit into memory. This is preferable to
+   * [[asMapSideInput]] if that's the case.
+   */
+  def asMapSingletonSideInput(implicit koder: Coder[K], voder: Coder[V]): SideInput[Map[K, V]] =
+    self
+      .transform(
+        _.groupByKey
+          .map { kv =>
+            require(kv._2.size == 1, s"Multiple values for key ${kv._1}")
+            (kv._1, kv._2.head)
+          }
+          .groupBy(_ => ())
+          .map(_._2.toMap)
```

Comment:
Could a `fold` work better in this case?

```scala
.map { kv =>
  require(kv._2.size == 1, s"Multiple values for key ${kv._1}")
  Map(kv._1 -> kv._2.head)
}
.fold
```

I'm not sure if `fold` would do preagg (and if that would make any real difference) but might be a bit more readable